File: packages/ruby/remote_process_client.rb
require 'socket'
require 'json'
require './model/game'
require './model/rules'

class RemoteProcessClient
  def initialize(host, port)
    @socket = TCPSocket::new(host, port)
    write_line('json')
  end
  def write_line(line)
    @socket.puts(line)
    @socket.flush
  end
  def write_token(token)
    write_line(token)
  end
  def read_game()
    if (line = @socket.gets).nil?
      return nil
    end
    json = JSON.parse(line)
    return Game.new(json)
  end
  def read_rules()
    if (line = @socket.gets).nil?
      return nil
    end
    json = JSON.parse(line)
    return Rules.new(json)
  end
  def write(actions)
    write_line(actions.to_json)
  end
end

File: packages/kotlin/src/main/kotlin/RemoteProcessClient.kt
import com.google.gson.Gson
import model.*

import java.io.*
import java.net.InetSocketAddress
import java.net.Socket
import java.nio.charset.StandardCharsets

class RemoteProcessClient @Throws(IOException::class)
constructor(host: String, port: Int) {
    internal var input: BufferedReader
    internal var output: OutputStreamWriter
    internal var gson = Gson()

    init {
        val socket = Socket()
        socket.tcpNoDelay = true
        socket.connect(InetSocketAddress(host, port))

        input = BufferedReader(
                InputStreamReader(BufferedInputStream(socket.getInputStream()), StandardCharsets.UTF_8))
        output = OutputStreamWriter(BufferedOutputStream(socket.getOutputStream()), StandardCharsets.UTF_8)

        output.write("json\n")
        output.flush()
    }

    @Throws(IOException::class)
    fun readGame(): Game? {
        val line = input.readLine()
        return gson.fromJson(line, Game::class.java)
    }

    @Throws(IOException::class)
    fun readRules(): Rules {
        val line = input.readLine()
        return gson.fromJson(line, Rules::class.java)
    }

    @Throws(IOException::class)
    fun write(actions: Map<Int, Action>) {
        val json = gson.toJson(actions)
        output.write(json)
        output.write("\n")
        output.flush()
    }

    @Throws(IOException::class)
    fun writeToken(token: String) {
        output.write(token)
        output.write("\n")
        output.flush()
    }
}